Logging: Introduce API for contextual logging

marefr opened the pull request:
**What this PR does / why we need it**:
Proposal to introduce an API for contextual logging to allow contextual key/value pairs to be attached, e.g. per request, so that any logger using this API will automatically get the per request context attached. The proposal makes the `traceID` available for contextual logger , if available, and would allow logs originating from a certain HTTP request to be correlated with `traceID`.  

A new `FromContext` method has been added to the logger interface to allow creating a contextual logger that has key/value pairs from `context.Context` attached:
https://github.com/grafana/grafana/blob/7d39eec4e2dd9382554ddeec299a8323ffd88e22/pkg/infra/log/interface.go#L15-L35

The tracing package registers a contextual log provider and this will be called whenever a logger is created from context (`Logger.FromContext(...)`):
https://github.com/grafana/grafana/blob/7d39eec4e2dd9382554ddeec299a8323ffd88e22/pkg/infra/tracing/tracing.go#L36-L42

Tried to call `tracing.TraceIDFromContext` from the log package, but it resulted in an import cycle. That's why my proposal includes support for registering (non-optimal) global contextual log providers.

In addition, don't attach `traceID` to context when it's empty and by that not including `traceID=00000000000000000000000000000000` in logs.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

marefr committed:
Logging: Introduce API for contextual logging

marefr committed:
add traceID to contextual logger for each HTTP request

marefr committed:
use contextual logger in various places

ying-jeanne commented on pkg/infra/log/log.go:
```diff
@@ -190,6 +191,32 @@ func (cl *ConcreteLogger) log(msg string, logLevel level.Value, args ...interfac
 }
 
+func (cl *ConcreteLogger) logCtx(ctx context.Context, msg string, logLevel level.Value, args ...interface{}) error {
+	allArgs := append([]interface{}{level.Key(), logLevel, "msg", msg}, args...)
```
It is nit, maybe we could enforce only allow log with context later. so we can search logs with trace id all the time.

marefr replied:
Can you elaborate on 

> enforce only allow log with context later

The `traceID` to be included in logs needs to always come from `context.Context` since it's generated on a per-HTTP-request.

ying-jeanne commented on pkg/infra/log/ctx_log.go:
```diff
@@ -0,0 +1,36 @@
+}
+
+// WithContextualArgs adds the key/value arguments to the provided context.
+func WithContextualArgs(ctx context.Context, args ...interface{}) context.Context {
```
actually the function is not using log, it takes a context and add the pair into the context, maybe we can leave it in the tracing package instead

marefr replied:
Related to below I guess?

> 1. Make [tracer.Start](https://github.com/grafana/grafana/blob/7700b529f951528e6fb8405b7fccba9fd01e26a1/pkg/infra/tracing/opentelemetry_tracing.go#L37) responsible for calling `log.WithContextualArgs(ctx, "traceID", traceID)`.

marefr commented:
Based on dicsussion with @bergquist he suggests a simpler API to make it easier to work with. Basically to inject a logger in `context.Context` that you can retrieve with `logger.FromContext(ctx)`. Something like this:

```go
package log
...
// FromContext returns a log.Logger previously injected into the given context.
// If no logger is present, a no-op logger is returned.
func FromContext(ctx context.Context) Logger {
	...
}

// ContextWithLogger injects the given logger into the given context
func ContextWithLogger(ctx context.Context, logger Logger) context.Context {
	...
}
```

We use named loggers in Grafana a lot. With above suggestion we would either skip providing logger name or manually include it after received the logger from context. Named loggers is very convenient when developing at least since you can turn on debug logging for specific loggers. 

If keeping logger names maybe something along these lines could work?

```go
type Logger interface {
	// existing methods ...

	// FromContext returns a new logger with contextual 
	FromContext(ctx context.Context) Logger
}

// example usage
myLogger := log.New("my-logger")
ctxLogger := myLogger.FromContext(ctx)
```

marefr committed:
change api to FromContext

marefr committed:
Merge remote-tracking branch 'origin/main' into poc_contextual_logger

marefr commented:
Changed to 

> ```go
> type Logger interface {
> 	// existing methods ...
> 
> 	// FromContext returns a new logger with contextual 
> 	FromContext(ctx context.Context) Logger
> }
> 
> // example usage
> myLogger := log.New("my-logger")
> ctxLogger := myLogger.FromContext(ctx)
> ```

marefr committed:
log package now accepts contextual log providers

marefr committed:
skip attaching zero trace id to context

When tracing not enabled, skip adding
traceID=00000000000000000000000000000000
to logs

marefr committed:
fix failing tests

wbrowne commented on pkg/util/proxyutil/reverse_proxy.go:
```diff
@@ -102,14 +102,16 @@ func errorHandler(logger glog.Logger) func(http.ResponseWriter, *http.Request, e
 			return
 		}
 
+		ctxLogger := logger.FromContext(r.Context())
```
Should line `100` not also use the `ctxLogger`?

marefr replied:
Good catch :+1: I'll fix

marefr committed:
fixes after review

joeblubaugh commented:
> `// If no logger is present, a no-op logger is returned.`

Would it be better to create a logger if there's nothing in the context? How can the caller tell if their logs are just going into the void?

marefr commented:
> > `// If no logger is present, a no-op logger is returned.`

@joeblubaugh Sorry, please disregard that comment and look at the actual changes instead.

Before these changes you would use `logger := log.New("<logger name>")` to create a new logger and log using that. With these changes by calling `logger.FromContext(ctx)` you'll get a new logger derived from earlier referenced logger.

wbrowne reviewed (approved):
LGTM

papagian commented on pkg/tests/api/alerting/api_alertmanager_test.go:
```diff
@@ -1075,7 +1075,6 @@ func TestAlertRuleCRUD(t *testing.T) {
 				require.NoError(t, err)
 
 				assert.Equal(t, res.Message, tc.expectedMessage)
-				assert.NotEmpty(t, res.TraceID)
```
why the `TraceID` no more required?

marefr replied:
Due to below and api tests doesn't have tracing enabled.

> In addition, don't attach `traceID` to context when it's empty and by that not including `traceID=00000000000000000000000000000000` in logs.

papagian reviewed (approved):
LGTM
